### PR TITLE
Go sdk preview docs

### PR DESCRIPTION
**Pull request by shykes**
* Update README in preparation for Go SDK preview
* Update copy in Go SDK introduction page, started by @vikram-dagger 
* Remove redundant "docs-0.3" docs, deprecated by Vikram's PR
* **NEW**: incorporate CUE and GO SDK into unified docs. *this is work in progress*

### CODE REVIEW

File: docs/v0.3/sdk/go/fssrz-index.md [hunk 1]
```diff
@@ -6,35 +6,68 @@ slug: /fssrz/sdk-go
...
+
+* A Go developer wishing their CI pipelines were Go instead of YAML
+* A developer who needs CI/CD, and is looking for an excuse to learn Go
+* Your team's "designated devops person", hoping to replace a pile of artisanal scripts with something more powerful
+* A platform engineer writing custom Go tooling, with the goal unifying continuous delivery across organizational silos
+* A cloud-native developer advocate or solutions engineer, looking to demonstrate a complex integration on short notice
+
+The Dagger GO SDK may *not* be a good fit if you are...
+
+* A developers who doesn't know Go, and is not interested in learning it
```
Comment (grouville) [inline docs/v0.3/sdk/go/fssrz-index.md:34]: a developer who doesn't

Comment (shykes) [reply docs/v0.3/sdk/go/fssrz-index.md]: Thanks! Fixed.

File: docs/new.sh [hunk 2]
```diff
@@ -10,7 +10,7 @@ function new() {
 		return 1
 	fi
 	local next
-	next=$(LC_ALL=C tr -dc 'a-z0-9' < /dev/urandom | head -c5)
+	next=$(LC_ALL=C tr -dc '0-9' < /dev/urandom | head -c6)
```
Comment (vikram-dagger) [inline docs/new.sh:13]: I'm curious about the reason for this change. Could you please add some context?

File: docs/v0.3/sdk/go/fssrz-index.md [hunk 3]
```diff
@@ -6,35 +6,68 @@ slug: /fssrz/sdk-go
...
-1. **Portable**. Your pipelines run in containers so you get the same behavior on your local machine and in your CI environment.
-2. **Scriptable**. Develop pipelines in Go, Typescript, Python, or even a shell script. No niche language or proprietary YAML is required.
-3. **Extensible**. Each pipeline has an API. Pipelines can be shared and reused across projects, teams or the entire community.
+The Dagger Go SDK contains everything you need to develop CI/CD pipelines in Go, and run them on any OCI-compatible container runtime.
 
-The Dagger Go SDK enables you to communicate with the Dagger API by calling native Go functions. With this SDK, you can author your CI/CD pipeline entirely in Go, without needing to learn a new language or think about interoperability.
+Currently the Go SDK consists of:
 
-## Usage
+* A go library: `dagger.io/dagger`
```
Comment (vikram-dagger) [inline docs/v0.3/sdk/go/fssrz-index.md:15]:
```suggestion
* A Go library: `dagger.io/dagger`
```

File: docs/v0.3/sdk/go/fssrz-index.md [hunk 4]
```diff
@@ -6,35 +6,68 @@ slug: /fssrz/sdk-go
...
+* This documentation
 
-The following sequence diagram explains how the Dagger SDK can be used in the typical CI/CD process flow:
+In practice, everything is optional except the Go library.
+
+## Who is it for?
+
+The Dagger Go SDK may be a good fit if you are...
+
+* A Go developer wishing their CI pipelines were Go instead of YAML
```
Comment (vikram-dagger) [inline docs/v0.3/sdk/go/fssrz-index.md:26]:
```suggestion
* A Go developer wishing your CI pipelines were Go instead of YAML
```

File: docs/v0.3/sdk/go/fssrz-index.md [hunk 5]
```diff
@@ -6,35 +6,68 @@ slug: /fssrz/sdk-go
...
-3. **Extensible**. Each pipeline has an API. Pipelines can be shared and reused across projects, teams or the entire community.
+The Dagger Go SDK contains everything you need to develop CI/CD pipelines in Go, and run them on any OCI-compatible container runtime.
 
-The Dagger Go SDK enables you to communicate with the Dagger API by calling native Go functions. With this SDK, you can author your CI/CD pipeline entirely in Go, without needing to learn a new language or think about interoperability.
+Currently the Go SDK consists of:
 
-## Usage
+* A go library: `dagger.io/dagger`
+* A CLI (optional)
+* Sample apps
```
Comment (vikram-dagger) [inline docs/v0.3/sdk/go/fssrz-index.md:17]:
```suggestion
* Sample applications
```

File: docs/v0.3/sdk/go/fssrz-index.md [hunk 6]
```diff
@@ -6,35 +6,68 @@ slug: /fssrz/sdk-go
...
+
+The Dagger Go SDK may be a good fit if you are...
+
+* A Go developer wishing their CI pipelines were Go instead of YAML
+* A developer who needs CI/CD, and is looking for an excuse to learn Go
+* Your team's "designated devops person", hoping to replace a pile of artisanal scripts with something more powerful
+* A platform engineer writing custom Go tooling, with the goal unifying continuous delivery across organizational silos
+* A cloud-native developer advocate or solutions engineer, looking to demonstrate a complex integration on short notice
+
+The Dagger GO SDK may *not* be a good fit if you are...
```
Comment (vikram-dagger) [inline docs/v0.3/sdk/go/fssrz-index.md:32]:
```suggestion
The Dagger Go SDK may *not* be a good fit if you are...
```

File: docs/v0.3/sdk/go/fssrz-index.md [hunk 7]
```diff
@@ -6,35 +6,68 @@ slug: /fssrz/sdk-go
...
+subgraph C["your deploy pipeline"]
   C1[" "] -.-> C2[" "] -.-> C3[" "] -.-> C4[" "]
 end
-goSDK -..-> A1 & B1 & C1
+
+lib -..-> engine -..-> oci -..-> A1 & B1 & C1
 ```
 
-## Next Steps
+1. Your Go program imports the Dagger go library
```
Comment (vikram-dagger) [inline docs/v0.3/sdk/go/fssrz-index.md:64]: Should we replace "go library" with "Go SDK" in this list?

File: docs/v0.3/sdk/go/fssrz-index.md [hunk 8]
```diff
@@ -6,35 +6,68 @@ slug: /fssrz/sdk-go
...
+2. Using the go library, your program opens a new session to a Dagger Engine: either by connecting to an existing engine, or by provisioning one on-the-fly.
+3. Using the go library, your program prepares API requests describing pipelines to run, then sends them to the engine. The wire protocol used to communicate with the engine is private and not yet documented, but this will change in the future. For now, the go library is the only documented API available to your program.
+4. When the engine receives an API request, it computes a DAG of low-level operations required to compute the result, and starts processing operations concurrently.
+5. When all operations in the pipeline have been resolved, the engine sends the pipeline result back to your program
```
Comment (vikram-dagger) [inline docs/v0.3/sdk/go/fssrz-index.md:68]:
```suggestion
5. When all operations in the pipeline have been resolved, the engine sends the pipeline result back to your program.
```

File: docs/v0.3/sdk/go/fssrz-index.md [hunk 9]
```diff
@@ -6,35 +6,68 @@ slug: /fssrz/sdk-go
...
+6. Your program may use the pipeline's result as input to new pipelines
```
Comment (vikram-dagger) [inline docs/v0.3/sdk/go/fssrz-index.md:69]:
```suggestion
6. Your program may use the pipeline's result as input to new pipelines.
```

File: docs/v0.3/sdk/go/8g34z-get-started.md [hunk 10]
```diff
@@ -6,10 +6,6 @@ slug: /8g34z/get-started-sdk-go
 
 ## Introduction
 
-[Dagger](https://dagger.io/) is a DevOps automation engine. It enables DevOps teams to orchestrate and deliver pipelines using any supported programming language. Popular use cases include multi-stage builds, CI/CD, on-demand staging environments, machine learning pipelines, end-to-end testing, artifact scanning and signing, and more.
-
-Dagger exposes its functionality via a low-level language-agnostic API framework and provides native language SDKs to communicate with this API. The Dagger Go SDK is one such SDK. It enables pipeline authors to use Dagger directly from Go, without needing to learn a new language.
-
 This tutorial teaches you the basics of using Dagger in Go. You will learn how to:
 
 - Install the Go SDK
```
Comment (vikram-dagger) [inline docs/v0.3/sdk/go/8g34z-get-started.md:11]: Should we replace "Go tool" in this guide with "Go program" for consistency with what is used in other docs in this PR?

File: docs/new.sh (hunk 2 above)
Comment (shykes) [reply docs/new.sh:13]: the non-numerical IDs look like garbled words,. my brain keeps trying to parse them which trips me up ever so slightly each time I glance at the url.

Basically: nitpicking :)

File: docs/v0.3/sdk/go/fssrz-index.md (hunk 7 above)
Comment (shykes) [reply docs/v0.3/sdk/go/fssrz-index.md:64]: - Until yesterday I would have said “yes”
- But yesterday @aluzzardi shared concern about us maybe conflating “library” and “sdk” to the point of maybe not matching the conventions of Go developers. This is an attempt at reconciling Andrea’s concern while continuing to use the SDK terminology. 

TLDR: it’s still a fine point under discussion but shouldn’t affect the overall structure.

**Comment by mircubed** [comment]
nitpicking, but the "Go SDK" and the "introduction" headers are the only ones that are clickable, so it feels weird since the other headers with the same weight aren't clickable.

**Comment by mircubed** [comment]
When I click on the preview, I get this page first....
Screen Shot 2022-10-19 at 9 33 22 AM

but then I click introduction and I see this...

Screen Shot 2022-10-19 at 9 35 45 AM


Not sure why the old introduction is coming up first in the preview, but worth checking out to make sure that the users don't see it when we merge.

File: docs/v0.3/sdk/go/8g34z-get-started.md (hunk 10 above)
Comment (shykes) [reply docs/v0.3/sdk/go/8g34z-get-started.md:11]: Do you mean remove “program” to write “tool” instead? Or the opposite?

**Comment by shykes** [comment]
> nitpicking, but the "Go SDK" and the "introduction" headers are the only ones that are clickable, so it feels weird since the other headers with the same weight aren't clickable.

Absolutely, that’s just me not having finished the implementation 👍

**Comment by shykes** [comment]
> When I click on the preview, I get this page first.... Screen Shot 2022-10-19 at 9 33 22 AM
> 
> but then I click introduction and I see this...
> 
> Screen Shot 2022-10-19 at 9 35 45 AM
> 
> Not sure why the old introduction is coming up first in the preview, but worth checking out to make sure that the users don't see it when we merge.

Yeah that’s leftover from original sidebar.

That part of the PR (cue sdk, merge sidebars) is not nearly as finished as the rest of the PR. Maybe I should separate but I didn’t want to over complicate right away. Wanted to get a general sense.

File: docs/v0.3/sdk/go/fssrz-index.md [hunk 11]
```diff
@@ -6,35 +6,68 @@ slug: /fssrz/sdk-go
...
+lib -..-> engine -..-> oci -..-> A1 & B1 & C1
 ```
 
-## Next Steps
+1. Your Go program imports the Dagger go library
+2. Using the go library, your program opens a new session to a Dagger Engine: either by connecting to an existing engine, or by provisioning one on-the-fly.
+3. Using the go library, your program prepares API requests describing pipelines to run, then sends them to the engine. The wire protocol used to communicate with the engine is private and not yet documented, but this will change in the future. For now, the go library is the only documented API available to your program.
+4. When the engine receives an API request, it computes a DAG of low-level operations required to compute the result, and starts processing operations concurrently.
+5. When all operations in the pipeline have been resolved, the engine sends the pipeline result back to your program.
+6. Your program may use the pipeline's result as input to new pipelines.
```
Comment (vito) [inline docs/v0.3/sdk/go/fssrz-index.md:69]: nitpick: Go capitalization

```suggestion
1. Your Go program imports the Dagger Go library
2. Using the Go library, your program opens a new session to a Dagger Engine: either by connecting to an existing engine, or by provisioning one on-the-fly.
3. Using the Go library, your program prepares API requests describing pipelines to run, then sends them to the engine. The wire protocol used to communicate with the engine is private and not yet documented, but this will change in the future. For now, the Go library is the only documented API available to your program.
4. When the engine receives an API request, it computes a DAG of low-level operations required to compute the result, and starts processing operations concurrently.
5. When all operations in the pipeline have been resolved, the engine sends the pipeline result back to your program.
6. Your program may use the pipeline's result as input to new pipelines.
```

**Review by vito — COMMENTED** [review]
I really like the "Get Started" page; it's awesome seeing a "CI/CD as a CLI" come together like that! The code needs to be updated to use the new `dagger.Connect()` API though. It'll make the code nicer too since you can get rid of the whole `engine.Start` + indented callback.

Context switching back to code for now. :)

**Comment by shykes** [comment]
Thanks all for the feedback, I have split up the contents from this PR into smaller, more focused ones.
